book/src/spec.md

```diff
@@ -0,0 +1,628 @@
+
+It is illegal to drive `endi` to the value \\(N\\) or greater.
+
+> This would semantically not be different from driving \\(N - 1\\), so there
+> is no reason to burden the sink logic by allowing this case.
+
+It is illegal to drive `endi` to a value less than `stai`.
+
+> This would essentially encode a VHDL null range, which, if allowed, could be
+> used to encode empty lists. However, detecting this condition is relatively
+> hard on FPGA resources (timing in particular), typically requiring a carry
+> chain for \\(N > 8\\). Instead, use `strb` for this purpose; at lower
+> complexities all `strb` bits must be equal, so only a single bit has to be
+> checked to determine whether a transfer carries any data.
+
+*[\\(C < 5\\)]* `endi` must be driven to \\(N - 1\\) by the source when `last`
+is zero, and may be ignored by the sink in this case.
+
+> Together with the other complexity rules up to this level, this means that
+> all lanes must be used for all but the last transfer containing elements for
+> the innermost list. This furthermore implies that the element with innermost
+> index \\(i\\) is always transferred on data lane \\(i \mod N\\).
+
+#### `strb` signal description
+
+The `strb` signal (strobe) is used to deactivate individual data lanes. It is
+a bit vector of size \\(N\\). If a `strb` bit is low, the associated data lane
+is deactivated.
+
+> Note that the opposite ("`strb` high -> activated) is *not* necessarily true
+> due to `endi` and `stai`.
+
+The state of the `strb` signal is significant only while `valid` is asserted.
+
+*[\\(C < 8\\)]* All `strb` bits must be driven to the same value by the source.
+The sink only needs to interpret one of the bits.
+
+> This effectively reduces the `strb` signal to a single bit indicating whether
+> the transfer is empty (low) or not (high), as it is illegal to deactivate all
+> data lanes by driving `endi` < `stai`.
+
+#### `user` signal description
+
+The `user` signal is the concatenation of a number of user-specified fields,
+carrying additional transfer-oriented information. The significance of this is
+user-defined.
+
+The state of the `user` signal is not significant when `valid` is not asserted.
+
+> The opposite is not necessarily true; it is up to the user's specifications
+> if (part of) the user signal can be insignificant despite `valid` being
+asserted.
+
+### Signal omission
+
+Not all signals are always needed. The following table shows the conditions for
+a signal to be relevant. If the condition is false, the signal may be omitted
+on the interface. When two interfaces with differing but compatible
+complexities are connected together, the default value specified in the table
+below must be driven for any omitted signals.
+
+| Name    | Condition                                   | Default   |
+|---------|---------------------------------------------|-----------|
+| `valid` | \\(\top\\)                                  | `'1'`     |
```

Comment:
It's clear enough to me but we could also leave out any signals that are never omitted from the table.

Reply:
`valid` and `ready` are in the table for their default value, because they can potentially be omitted for sources that are always valid (such as a /dev/zero-like source, some status signal sampled at max sink speed, etc) and sinks that are always ready (such as a /dev/null-like source, something like a counter, etc.). Xilinx' interpretation of AXI4-stream IP has these defaults too. I added this below the table; if you guys disagree with it let me know.